bfasst/ninja_flows/flow_utils.py

```diff
@@ -30,3 +32,9 @@ def create_build_file():
     """Overwrite the build.ninja file when running generation and build steps separately"""
     with open(NINJA_BUILD_PATH, "w") as f:
         f.write("# This file is autogenerated by bfasst\n\n")
+
+
+def configure_vivado_synth_tool(design, flow_args, ooc=False):
+    if isinstance(flow_args, dict) and "synth" in flow_args:
+        return VivadoSynth(design, flow_args["synth"], ooc)
+    return VivadoSynth(design, ooc=ooc)
```

Comment:
It seems this function is just checking if `"synth"` is in `flow_args`?

We could entirely bypass the need for this function by making sure `flow_args` was a [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict)

Then, anywhere we are calling this function, we could simply call 

```python
VivadoSynth(design, flow_args["synth"], ooc)
```

And it would work regardless of whether "synth" was a key in the dict or not.  

I think this would be more readable than calling into a separate function to create the object.

Reply:
To do this, would I go to the place where flow_args is initialized and make it a defaultdict instead of a dict? And then from there I could replace everywhere configure_vivado_synth_tool(design, flow_args, ooc) is called with VivadoSynth(design, flow_args["synth"], ooc)?

Reply:
@yonnorc42 

Yes, exactly.

Reply:
Also, you could use flow_args.get("synth"). According to the [docs](https://docs.python.org/3.10/library/stdtypes.html?highlight=dict#mapping-types-dict): the `dict.get(key)` will return None if the key does not exist, which is the default value for the flow_args parameter in the VivadoSynth constructor. I think this may be better instead of a default dict? Also @KeenanRileyFaulkner, I see there is an `isinstance()` check on flow args - is there ever a case where it is not a dict? 
I think we could just replace every call to `configure_vivado_synth_tool` with `VivadoSynth(design, flow_args.get("synth"), ooc=ooc)`

Reply:
My understanding is the defaultdict woult prevent a KeyError from ever being thrown on a bad access. I'm not sure if we want to totally remove that functionality, in my opinion the use of dict.get() ensures that in this one place we are okay with it returning None instead of throwing a KeyError on a missing key.

Reply:
I agree.